File: backend/models/Announcement.js
const mongoose = require('mongoose');

const announcementSchema = new mongoose.Schema({
  title: {
    type: String,
    required: [true, 'Title is required'],
    trim: true,
    maxlength: [200, 'Title cannot exceed 200 characters']
  },
  content: {
    type: String,
    required: [true, 'Content is required'],
    maxlength: [5000, 'Content cannot exceed 5000 characters']
  },
  author: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  category: {
    type: String,
    enum: ['general', 'academic', 'event', 'urgent', 'notice'],
    default: 'general'
  },
  priority: {
    type: String,
    enum: ['low', 'medium', 'high', 'critical'],
    default: 'medium'
  },
  targetAudience: [{
    type: String,
    enum: ['all', 'students', 'faculty', 'admin', 'specific-dept']
  }],
  department: {
    type: String,
    trim: true
  },
  attachments: [{
    filename: String,
    url: String,
    size: Number,
    type: String
  }],
  isPublished: {
    type: Boolean,
    default: false
  },
  publishedAt: {
    type: Date
  },
  expiresAt: {
    type: Date
  },
  viewers: [{
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  }], // Track who has viewed the announcement
  likes: [{
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  }],
  comments: [{
    user: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true
    },
    content: {
      type: String,
      required: true
    },
    createdAt: {
      type: Date,
      default: Date.now
    }
  }]
}, {
  timestamps: true
});

// Index for efficient querying
announcementSchema.index({ createdAt: -1 });
announcementSchema.index({ category: 1, createdAt: -1 });
announcementSchema.index({ targetAudience: 1, createdAt: -1 });
announcementSchema.index({ expiresAt: 1 }, { expireAfterSeconds: 0 });

module.exports = mongoose.model('Announcement', announcementSchema);

File: frontend/cec_echo_mobile/lib/models/announcement.dart
import 'package:cec_echo_mobile/models/user.dart';

User _fallbackUser(String id) {
  return User(
    id: id,
    username: 'user',
    email: '',
    firstName: 'User',
    lastName: '',
    role: 'student',
    isActive: true,
  );
}

class Announcement {
  final String id;
  final String title;
  final String content;
  final User author;
  final String category;
  final String priority;
  final List<String> targetAudience;
  final String? department;
  final List<dynamic>? attachments;
  final bool isPublished;
  final DateTime? publishedAt;
  final DateTime? expiresAt;
  final List<User> viewers;
  final List<User> likes;
  final List<Comment> comments;
  final DateTime createdAt;
  final DateTime updatedAt;

  Announcement({
    required this.id,
    required this.title,
    required this.content,
    required this.author,
    required this.category,
    required this.priority,
    required this.targetAudience,
    this.department,
    this.attachments,
    required this.isPublished,
    this.publishedAt,
    this.expiresAt,
    required this.viewers,
    required this.likes,
    required this.comments,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Announcement.fromJson(Map<String, dynamic> json) {
    final authorRaw = json['author'];
    final viewerRaw = json['viewers'] as List<dynamic>? ?? [];
    final likesRaw = json['likes'] as List<dynamic>? ?? [];
    final commentsRaw = json['comments'] as List<dynamic>? ?? [];

    return Announcement(
      id: json['_id'] ?? json['id'],
      title: json['title'],
      content: json['content'],
      author: authorRaw is Map<String, dynamic>
          ? User.fromJson(authorRaw)
          : _fallbackUser(authorRaw?.toString() ?? ''),
      category: (json['category'] ?? 'general').toString(),
      priority: (json['priority'] ?? 'medium').toString(),
      targetAudience: List<String>.from(json['targetAudience'] ?? const ['all']),
      department: json['department'],
      attachments: json['attachments'],
      isPublished: json['isPublished'] ?? true,
      publishedAt: json['publishedAt'] != null ? DateTime.parse(json['publishedAt']) : null,
      expiresAt: json['expiresAt'] != null ? DateTime.parse(json['expiresAt']) : null,
      viewers: viewerRaw
          .map((user) => user is Map<String, dynamic> ? User.fromJson(user) : _fallbackUser(user.toString()))
          .toList(),
      likes: likesRaw
          .map((user) => user is Map<String, dynamic> ? User.fromJson(user) : _fallbackUser(user.toString()))
          .toList(),
      comments: commentsRaw
          .whereType<Map<String, dynamic>>()
          .map(Comment.fromJson)
          .toList(),
      createdAt: DateTime.parse(json['createdAt']),
      updatedAt: DateTime.parse(json['updatedAt']),
    );
  }
}

class Comment {
  final String id;
  final User user;
  final String content;
  final DateTime createdAt;

  Comment({
    required this.id,
    required this.user,
    required this.content,
    required this.createdAt,
  });

  factory Comment.fromJson(Map<String, dynamic> json) {
    final userRaw = json['user'];
    return Comment(
      id: json['_id'] ?? json['id'],
      user: userRaw is Map<String, dynamic>
          ? User.fromJson(userRaw)
          : _fallbackUser(userRaw?.toString() ?? ''),
      content: json['content'],
      createdAt: DateTime.parse(json['createdAt']),
    );
  }
}
